be/src/olap/schema_change.cpp

```diff
@@ -1490,133 +1493,131 @@ OLAPStatus SchemaChangeHandler::_do_process_alter_tablet_v2(const TAlterTabletRe
-                                  end_version);
-        if (res != OLAP_SUCCESS) {
-            LOG(WARNING) << "init delete handler failed. base_tablet=" << base_tablet->full_name()
-                         << ", end_version=" << end_version;
+            // init one delete handler
+            int32_t end_version = -1;
+            for (auto& version : versions_to_be_changed) {
+                if (version.second > end_version) {
+                    end_version = version.second;
+                }
+            }
 
-            // release delete handlers which have been inited successfully.
-            delete_handler.finalize();
-            break;
-        }
+            res = delete_handler.init(base_tablet->tablet_schema(), base_tablet->delete_predicates(),
+                                      end_version);
+            if (res != OLAP_SUCCESS) {
+                LOG(WARNING) << "init delete handler failed. base_tablet=" << base_tablet->full_name()
+                             << ", end_version=" << end_version;
 
-        // acquire data sources correspond to history versions
-        base_tablet->capture_rs_readers(versions_to_be_changed, &rs_readers, mem_tracker);
-        if (rs_readers.size() < 1) {
-            LOG(WARNING) << "fail to acquire all data sources. "
-                         << "version_num=" << versions_to_be_changed.size()
-                         << ", data_source_num=" << rs_readers.size();
-            res = OLAP_ERR_ALTER_DELTA_DOES_NOT_EXISTS;
-            break;
-        }
+                // release delete handlers which have been inited successfully.
+                delete_handler.finalize();
+                break;
+            }
 
-        for (auto& rs_reader : rs_readers) {
-            res = rs_reader->init(&reader_context);
-            if (res != OLAP_SUCCESS) {
-                LOG(WARNING) << "failed to init rowset reader: " << base_tablet->full_name();
+            // acquire data sources correspond to history versions
+            base_tablet->capture_rs_readers(versions_to_be_changed, &rs_readers, mem_tracker);
+            if (rs_readers.size() < 1) {
+                LOG(WARNING) << "fail to acquire all data sources. "
+                             << "version_num=" << versions_to_be_changed.size()
+                             << ", data_source_num=" << rs_readers.size();
+                res = OLAP_ERR_ALTER_DELTA_DOES_NOT_EXISTS;
                 break;
             }
-        }
 
-    } while (0);
+            for (auto& rs_reader : rs_readers) {
+                res = rs_reader->init(&reader_context);
+                if (res != OLAP_SUCCESS) {
+                    LOG(WARNING) << "failed to init rowset reader: " << base_tablet->full_name();
+                    break;
+                }
+            }
 
-    new_tablet->release_header_lock();
-    base_tablet->release_header_lock();
+        } while (0);
+    }
     new_tablet->release_push_lock();
```

Comment:
I think we could reference clickhouse's  scope_safe_guard.h， then we could use RAII mode to manage the lock's lock or unlock. And will not forget to unlock.
There are many code like this in doris:

acquire lock xxxx
do something xxxx
release lock xxxx